### PR TITLE
Manual: New chapter on polymorphic types and functions in the tutorial

### PR DESCRIPTION
**Pull request by Octachron**
This PR proposes to add a new chapter to the manual to describe in more detail the properties of polymorphic functions and type in OCaml. More precisely, the new chapter would describe:
- the relaxed value restriction: starting from weakly polymorphic types to the relaxed value restriction itself
- polymorphic recursion and the use of explicit polymorphic annotation
- higher-rank polymorphic functions using explicitly polymorphic record fields or object methods.

The main objective of this PR is to gather in the manual the information that is already available on the web but not always easily discoverable.

As an important side-effect, this PR moves the description of explicitly polymorphic type annotation out of the extension section of the language reference manual.

Note: this is still a work in progress, comments and criticisms are more than welcome.

### CODE REVIEW

File: manual/manual/tutorials/polymorphism.etex
Comment (objmagic) [inline manual/manual/tutorials/polymorphism.etex]: there are three main ways

File: manual/manual/tutorials/polymorphism.etex
Comment (objmagic) [inline manual/manual/tutorials/polymorphism.etex]: swaps

File: manual/manual/tutorials/polymorphism.etex
Comment (objmagic) [inline manual/manual/tutorials/polymorphism.etex]: with

File: manual/manual/tutorials/polymorphism.etex
Comment (objmagic) [inline manual/manual/tutorials/polymorphism.etex]: abstract

Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: Thanks for noticing these errors.

File: manual/manual/tutorials/polymorphism.etex
Comment (gasche) [inline manual/manual/tutorials/polymorphism.etex]: mutation

File: manual/manual/tutorials/polymorphism.etex
Comment (gasche) [inline manual/manual/tutorials/polymorphism.etex]: You _could_ point out that this would be problematic if two distinct modules relied on this module's interface, unifying exported weak types to different things. You _should_ spell out the error message in entirety (for example, by using an example that raises it), because it's confusing when you first see it and it would be a good anchor for people looking for the error message text.

File: manual/manual/tutorials/polymorphism.etex
Comment (gasche) [inline manual/manual/tutorials/polymorphism.etex]: People sometimes use "positive position" instead of "covariant position", and it makes as much sense when subtyping is not considered. It would be helpful to refer to this other wording at some point.

File: manual/manual/tutorials/polymorphism.etex
Comment (gasche) [inline manual/manual/tutorials/polymorphism.etex]: I feel that this description is too abstract. (preserves the order of the subtyping relationship?).

File: manual/manual/tutorials/polymorphism.etex
Comment (gasche) [inline manual/manual/tutorials/polymorphism.etex]: Another explanation would be that as soon as a type variable appears in a position describing mutable state, such as `'a ref`, it becomes invariant, and that thus covariant variables will never denote mutable locations and can be safely generalized.

I don't know whether we need to have both explanations. I feel that you are repeating a lot of what was rather clearly written in Jacques Garrigue's article, and that it may be enough to more concisely explain the behaviour and intuition, and refer to the article for people looking for an in-depth explanation.

(Also that could help in keeping the documentation less abstract/formal.)

File: manual/manual/tutorials/polymorphism.etex
Comment (gasche) [inline manual/manual/tutorials/polymorphism.etex]: That code is too complex; what about just using lists under the hood? (Or `int -> 'a option` if you want something exotic)

File: manual/manual/tutorials/polymorphism.etex
Comment (gasche) [inline manual/manual/tutorials/polymorphism.etex]: Let's avoid "associated type" which means something else. You could just say "When the type definitions are exposed".

**Comment by gasche** [comment]
I think that's a good idea. I reviewed the part before "Polymorphic recursion" so far. I think it is important to keep things simple, concrete and example-based as soon as we can, and avoid formalism -- because it can turn people away and make the document feel less accessible.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: Fixed.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: Good points. I have also added a remark on the fact that this error cannot be repoduced in the toplevel interpreter to avoid potential confusion.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: I have added a short mention. It may be better to weave the name more cohesively in the text.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: You are right, I fear that I let out too much of my inner mathematician here. Anyway, the whole section have been rewritten (see below).

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: I have tentatively rewritten the variance section to put more emphasis on concrete examples. Linking directly invariance with mutable state seems to give a better flow to the text.
I would also add a link to Jacques Guarrigue's article later.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: Sure.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: Fixed.

**Comment by Octachron** [comment]
@gasche 
Thanks for your time and review.

**Comment by bluddy** [comment]
Thank you so much for doing this!

File: manual/manual/tutorials/polymorphism.etex
Comment (gasche) [inline manual/manual/tutorials/polymorphism.etex]: `fixing` suggests that something was "broken" before, you should rather use something like

> To solve this error, it suffices to add an explicit type annotation to make the type precise at declaration time.

I would also add:

> This is in any case a good practice for such global mutable variables; otherwise, it will "adopt" the first type of use, which can give confusing type errors if there is a mistake at the first use site -- later, correct uses would be flagged as errors.

File: manual/manual/tutorials/polymorphism.etex
Comment (gasche) [inline manual/manual/tutorials/polymorphism.etex]: I'm not sure it makes sense to show the two solutions on the same example. It seems pretty clear when I read `let foo = ref None` that a mutable global variable is intended to be initialized later, and making it a function changes its semantics in a way that seems rather artificial to me. Later examples will propose eta-expansion as a solution to get more polymorphism, so maybe you could just this ", or...." part of the paragraph.

File: manual/manual/tutorials/polymorphism.etex
Comment (gasche) [inline manual/manual/tutorials/polymorphism.etex]: I think we could skip that paragraph as well -- I'm surprised I notice how to shorten other people's texts when mine are always so verbose. The explanation you give is not very good¹, and in any case people are rarely surprised when stuff works, they want explanations when it does not.

¹: The reason why weak polymorphic values are forbidden at module interfaces is that the module may be linked twice with different unifications, making it unsound. On the other hand, a toplevel declaration cannot be used by two consumers simultaneously, there is a single "continuation" which is the rest of the toplevel section. The sentence below ("in the toplevel, ...") is not really an explanation.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: Good points. I have adopted your new paragraph with some stylistic modification.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: Yes, skipping the second point is probably a good idea since it does not really make sense with this specific example.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: You are right the explanation is bad.

Anyway, I was worried about people trying to reproduce the error message in a toplevel after reading the previous paragraph. But it is probably a too unfrequent occurence to warrant any explanation.

**Comment by damiendoligez** [comment]
@objmagic @gasche Are you done reviewing?

**Comment by gasche** [comment]
I'm not done, unfortunately. I need to invest more time and there are in places some reformulation that I'm thinking of proposing (instead of just commenting). Are you asking because you would like to merge soon? What's the calendar for when this should be settled?

(This manual change is more ambitious than others and it also hits closer to home in terms of discussing non-trivial points of advanced type system / language design. I could see myself falling into an unbounded process of gradual refinement (especially if we get Didier involved, which would be a good idea). I think that we should instead aim for a mergeable version that is reasonable, and refine later if we want to -- le mieux est l'ennemi du bien -- but I haven't read the whole thing yet so I cannot say I'm there.)

@Octachron what's your own feeling/preference? Does it feel "ready" to you, do you have specific expectations in term of perfectionism level vs. merge time?

**Comment by Octachron** [comment]
For this specific PR, my objective is more to reach a good first step that could be readily improved upon in the future, either directly or by collecting feedback. I readily admit that with the limited depth of my understanding on some of the subjects covered trying to get it right on the first version would be overly ambitious.

However, there are still rough parts, spelling errors and some unnecessary complexity
in the current version, that I would like to correct (by the end of the week if needed?). I also would not mind some external confirmation that there is no factual errors in the later part of the text.

@gasche, if you find the time to write these propositions, I would be interested to merge them.

p.s. : 

@damiendoligez , @gasche: As an echo to the trunk manual discussion, this is another situation where I feel that it might be nice to have an experimental web version of the manual to gather feedback more easily.

**Comment by damiendoligez** [comment]
@gasche There's no calendar, I was just pinging the PR to see if it was ready.

File: manual/manual/tutorials/polymorphism.etex
Comment (gerdstolpmann) [inline manual/manual/tutorials/polymorphism.etex]: Give an example here, e.g.

```
let store = ref None
  store : `_a option ref
store := Some 1
  store : int option ref
```

File: manual/manual/tutorials/polymorphism.etex
Comment (gerdstolpmann) [inline manual/manual/tutorials/polymorphism.etex]: values in toplevel modules

File: manual/manual/tutorials/polymorphism.etex
Comment (gerdstolpmann) [inline manual/manual/tutorials/polymorphism.etex]: Stumbling as reader over this. Better: "this rule is not only impractical but also not even a sufficient solution to the problem, as illustrated ..."

File: manual/manual/tutorials/polymorphism.etex
Comment (gerdstolpmann) [inline manual/manual/tutorials/polymorphism.etex:163]: Avoid variance in the main text, e.g. "In simple words, the generalization can also be done when the type only refers to values returned by functions that never appear as arguments of other functions (i.e. the values are all in covariant positions)"

File: manual/manual/tutorials/polymorphism.etex
Comment (gerdstolpmann) [inline manual/manual/tutorials/polymorphism.etex]: typo: genericity

File: manual/manual/tutorials/polymorphism.etex
Comment (gerdstolpmann) [inline manual/manual/tutorials/polymorphism.etex]: Before going on with this "brittle" stuff, show the reader safe grounds. Explain what a regular ADT is before going on with non-regular types. It is important to see the difference.

File: manual/manual/tutorials/polymorphism.etex
Comment (gerdstolpmann) [inline manual/manual/tutorials/polymorphism.etex]: I am missing here two "footnotes". First, the practical comment that many types in explicitly annotated functions can be simply left as "_", e.g. `let rec depth : 'a . 'a nested -> _`, limiting the annotation effort.

Second, point out that there is also a variation using "newtypes" (`let rec depth : type t . t nested -> _`) which often yields the same effect though the treatment of the variable inside the function is slightly different when GADTs are involved (i.e. the variable can be specialized differently depending on the GADT case whereas the normal type variable `'a` means always the same type on the current recursion level).

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: Thanks for the suggestion, giving an example here improve nicely the flow of the section.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: I would rather say that weak types cannot appear in signatures in order
to include cases like

```
module M = struct module N = struct let store = ref None end end
```

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: Thanks for your remark, I did not notice that this hand-wavy description fits multiple
type system (some which are even sound as far as I can tell). So it is probably better
to delete my babbling and shows the problem directly. For instance, 

> Identifying the exact context in which polymorphic types should be
> replaced by weak types in a modular way is a difficult question. Indeed
> the type system system must take in account the fact that functions
> can hide persistent mutable state, as illustrated ...
> At the same time, it should be possible to use a local mutable state without
> impacting the type of a function.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex:163]: I am not sure that it is better to avoid the word variance here. I fear that any reasonable approximation would end up quite wordy:

> "As a first approximation, the generalization can also be done when the type only refers to values returned by functions that never appear as arguments of other functions or parameter of a mutable type constructor."

Is it not simpler to use the name "variance" and then explains what is variance? Note that it may be well my academic background speaking here.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: Well, up to now, I was not fully aware of the distinction between regular and non-regular ADT. Thanks for the pointer! And you are right making the distinction will make the origin of the problem much clearer for the reader.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [inline manual/manual/tutorials/polymorphism.etex]: > First, the practical comment that many types in explicitly annotated functions can be simply left as "_", e.g. let rec depth : 'a . 'a nested -> _, limiting the annotation effort.

Good point.

> Second, point out that there is also a variation using "newtypes" (let rec depth : type t . t nested -> _) which often yields the same effect though the treatment of the variable inside the function is slightly different when GADTs are involved (i.e. the variable can be specialized differently depending on the GADT case whereas the normal type variable 'a means always the same type on the current recursion level).

It is certainly an important point to make and this is the right place to do so, but I fear this not the right time since GADT and locally abstract type are still relegated to the extension section of the manual.

**Comment by Octachron** [comment]
I have updated the chapter to take in account @gerdstolpmann comments; thanks again.
As a side note, I have also an [experimental html page](http://www.polychoron.fr/ocaml-nonmanual/polymorphism.html) for the chapter,  for potentially interested readers that were detered by the raw latex (and extensions) files.

**Comment by nojb** [comment]
The page looks great, except that the table of contents overlaps with the main content.  Using Safari on OS X 10.11.

**Comment by Octachron** [comment]
I have just fixed this specific rendering issue on chrome. I am not completely sure if it works now with Safari.

**Comment by nojb** [comment]
Yes, it works.  Thanks!

**Comment by alainfrisch** [comment]
I think we should try to merge this nice contribution before it's forgotten.

@gasche What's your opinion?  Do you think we can merge in the current form?  The text can always be improved later on.

File: manual/manual/tutorials/polymorphism.etex
Comment (damiendoligez) [inline manual/manual/tutorials/polymorphism.etex]: Missing word: "**is** known"

File: manual/manual/tutorials/polymorphism.etex
Comment (damiendoligez) [inline manual/manual/tutorials/polymorphism.etex]: Why "2"? Should be "3" no?

File: manual/manual/tutorials/polymorphism.etex
Comment (damiendoligez) [inline manual/manual/tutorials/polymorphism.etex]: Don't you mean "supertype" here?

File: manual/manual/tutorials/polymorphism.etex
Comment (damiendoligez) [inline manual/manual/tutorials/polymorphism.etex]: Jacques Garrigue

**Comment by gasche** [comment]
@alainfrisch unfortunately I'm not sure I will have any time to review the rest of this PR before the release. I would be happy to trust you (and/or @damiendoligez, etc.) and merge if you are happy with the current state.

**Comment by alainfrisch** [comment]
I've looked only superficially, so I'll @damiendoligez decides.

**Review by damiendoligez — CHANGES_REQUESTED** [review]
It'll be good to merge when my few comments (above) are addressed.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [reply manual/manual/tutorials/polymorphism.etex]: Fixed.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [reply manual/manual/tutorials/polymorphism.etex]: Indeed.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [reply manual/manual/tutorials/polymorphism.etex]: There was indeed an order inversion problem here. I choose to replace it by `"x" a subtype of "xy"` to avoid introducing the notion of `supertype`.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [reply manual/manual/tutorials/polymorphism.etex]: Fixed.

**Comment by Octachron** [comment]
I have fixed the issues raised by @damiendoligez , corrected few grammatical errors or idiosyncrasies along the way and added a Changes entry.

File: manual/manual/tutorials/polymorphism.etex
Comment (garrigue) [reply manual/manual/tutorials/polymorphism.etex]: I suppose he's using OCaml cast syntax here: `(_ : xy :> x)`. Might require some explanation.

File: manual/manual/tutorials/polymorphism.etex
Comment (garrigue) [inline manual/manual/tutorials/polymorphism.etex]: That's the canonical one.

File: manual/manual/tutorials/polymorphism.etex
Comment (garrigue) [inline manual/manual/tutorials/polymorphism.etex]: `There are two major differences between`

**Review by damiendoligez — APPROVED** [review]

**Comment by damiendoligez** [comment]
Good to go, and we can always enhance it later.

Should we squash the commits before merging?

**Comment by Octachron** [comment]
I think that squashing the commits is probably better for this PR, since the majority of commits after the first one are small fixes with an useless commit message for an external reader.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [reply manual/manual/tutorials/polymorphism.etex]: Fixed, thanks.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [reply manual/manual/tutorials/polymorphism.etex]: Thanks for confirming.

File: manual/manual/tutorials/polymorphism.etex
Comment (Octachron) [reply manual/manual/tutorials/polymorphism.etex]: I am indeed using OCaml cast syntax. Thinking about it, it may be better to only use words, i.e. `is a subtype` rather than symbolic notation.

**Review by gasche — APPROVED** [review]
Approved (on the basis of Damien's more exhaustive review). @Octachron, would you do the work of squashing, merging and cherry-picking in 4.06?

**Comment by Octachron** [comment]
@gasche : Certainly, I will.

**Comment by Octachron** [comment]
Merged and cherry-picked on 4.06(641dab319e7b3f75a).

**Comment by gasche** [comment]
Thanks! And thank for your work on the manual in general.

Something I've been thinking about but haven't got around to doing it yet is to make a pre-release of the 4.06 manual in a temporary place, to include in 4.06-preparation-related announcements. If you also think this could be of use, would you be interested in taking care of it? (I have web space where it's easy to push temporary stuff; if you don't, I will handle the whole thing).

**Comment by Octachron** [comment]
I think releasing a beta version of the manual is a good idea. If it just a matter of putting a 4.06 version of the manual online, I could do the same thing as http://www.polychoron.fr/ocaml-nonmanual/mpr7497/ and make it available at http://www.polychoron.fr/ocaml-beta-manual/4.06/ with a slightly adjusted disclaimer text (and keeping the "alternative" color scheme).

**Comment by gasche** [comment]
Yes, I think that would be very nice. You could advertise it in the email thread of the latest beta announcement, and possibly point to specific part of the manuals that changed (typically this new section).

**Comment by Drup** [comment]
@gasche Couldn't we just have a version of the manual for trunk automatically uploaded ? (http://caml.inria.fr/pub/docs/manual-ocaml-trunk/ , for example)

**Comment by gasche** [comment]
We certainly could, but someone has to do the work.

**Pull request merged**
